File: Python/UnionFind.py
class UnionFind:
    def __init__(self, n: int) -> None: self.__par = [-1] * n
    def __len__(self) -> int: return len(self.__par)
    def root(self, i: int) -> int:
        if self.__par[i] >= 0:
            self.__par[i] = self.root(self.__par[i])
            return self.__par[i]
        return i
    def size(self, i: int) -> int: return -self.__par[self.root(i)]
    def unite(self, i: int, j: int) -> bool:
        i = self.root(i)
        j = self.root(j)
        if i == j: return False
        if i > j: i, j = j, i
        self.__par[i] += self.__par[j]
        self.__par[j] = i
        return True
    def same(self, i: int, j: int) -> bool: return self.root(i) == self.root(j)
    def groups(self) -> list[list[int]]:
        n = len(self.__par)
        res = [[] for _ in range(n)]
        for i in range(n): res[self.root(i)].append(i)
        return list(filter(lambda x: x, res))

def is_bipartite(uf: UnionFind) -> bool:
    assert len(uf) % 2 == 0
    n = len(uf) // 2
    ok = True
    for i in range(n): ok &= not uf.same(i, i + n)
    return ok

File: Nim/UnionFind.nim
import sequtils
type UnionFind = ref object
    par: seq[int]
proc initUnionFind*(n: int): UnionFind {.inline} = UnionFind(par: newSeqWith(n, -1))
proc `[]`*(uf: UnionFind, k: int): int {.inline} =
    var i = k
    while uf.par[i] >= 0:
        let p = uf.par[uf.par[i]]
        if p < 0: return uf.par[i]
        uf.par[i] = p
        i = p
    i
proc size*(uf: UnionFind, i: int): int {.inline} = -uf.par[uf[i]]
proc unite*(uf: UnionFind, i: int, j: int) {.inline} =
    var x = uf[i]
    var y = uf[j]
    if x == y: return
    if x > y: swap(x, y)
    uf.par[x] += uf.par[y]
    uf.par[y] = x
proc groups*(uf: UnionFind): seq[seq[int]] {.inline} =
    let n = uf.par.len
    var res = newSeq[seq[int]](n)
    for i in 0..<n: res[uf[i]].add(i)
    res.filterIt(it.len > 0)

proc isBipartite*(uf: UnionFind): bool {.inline} =
    assert(uf.par.len mod 2 == 0)
    let n = uf.par.len div 2
    var ok = true
    for i in 0..<n: ok = ok and uf[i] != uf[i + n]
    ok